docs: post how does sync schema work in Bytebase technical blog (#656)
bytebase/bytebase.com: content/blog/how-does-sync-schema-work-in-Bytebase.md

---

> title: How does sync schema work in Bytebase

Reviewer: How schema sync works in Bytebase

how-schema-sync-work

author: Zipeng
published_at: 2022/11/27 15:40:00
feature_image: /static/blog/how-does-sync-schema-work-in-Bytebase/feature.webp
tags: Education

> description: Technical contents of Bytebase sync schema for MySQL.

Reviewer: ```suggestion
description: Technical details about how Bytebase implements schema synchronization for MySQL.
```

---

> ## What is schema synchronization?

Reviewer: ```suggestion
## What is Schema Synchronization?
```

> Schema synchronization can calculate the difference between two databases and give the SQLs, which can migrate to let the schema be the same as the destination database. People do not need to write the differences by hand; the machine does it.

Reviewer: ```suggestion
Schema synchronization can calculate the difference between two database schemas, and generate the SQL statements representing the diffs. Thus, people do not need to write the migration SQL statements by hand.
```

## Overview

> Bytebase has supported MySQL schema synchronization since v1.8.0. The data pipeline for this module in Bytebase is shown as follows.

Reviewer: ```suggestion
Bytebase has supported MySQL schema synchronization since v1.8.0. Below is the workflow.
```

![schema-diff-data-flow](/static/blog/how-does-sync-schema-work-in-Bytebase/schema-diff-data-flow.webp)

## Schema Diff Engine

### SQL AST

> Bytebase can dump the database schema and process them into [abstract syntax tree (AST)](https://en.m.wikipedia.org/wiki/Abstract_syntax_tree) via a parser.

Reviewer: ```suggestion
Bytebase first dumps the database schema and converts them into [abstract syntax tree (AST)](https://en.m.wikipedia.org/wiki/Abstract_syntax_tree) via the parser.
```

![SQL-AST](/static/blog/how-does-sync-schema-work-in-Bytebase/sql-ast.webp)

### Diff on SQL AST

> We can get object properties in AST. Compare two corresponding entities (which may not exist). The following chart shows a simple example:

Reviewer: ```suggestion
Bytebase then walks through the ASTs for both database schemas and compares two corresponding entities (which may not exist) from ASTs. The following chart shows a simple example:
```

![diff-on-sql-ast](/static/blog/how-does-sync-schema-work-in-Bytebase/diff-on-sql-ast.webp)

> ### Safe Order

Reviewer: ```suggestion
### Handle Dependencies
```

Objects and operations in a database have dependencies, such as the need to ensure that a table exists before adding a column. The following subsections show how Bytebase handles these dependencies.

#### Dependencies between different objects

> There are many dependencies between database objects, like constraints depending on indices and columns, and columns depending on the table. We need to deal with them according to the topological order. The adding order must follow  the below list, and deleting order should follow the reverse order:

Reviewer: ```suggestion
There are many dependencies between database objects, like constraints depending on indices and columns, and columns depending on the table. We need to deal with them according to the topological order. The creating order must follow the below list, and the deleting order should follow the reverse order:
```

1. Table
2. Column
3. View
4. Index
5. Constraint

> Function, trigger, and procedure do not validate at creation time, so we can create them anywhere after deleting them.

Reviewer: ```suggestion
Function, trigger, and procedure are not validated at creation time, so we can create them anywhere after deleting them.
```

Reviewer: I don't get this sentence BTW.

What does "them" refer to?

What is "create after deleting"?

Author: How about now?

#### Dependencies between different operation types

> Also, we should focus on the dependencies between operation types. The below list shows the ordering:

Reviewer: ```suggestion
Also, we should sort dependencies among operation types. Below lists the ordering:
```

> 1. Additions for new nodes, like adding a new column.

Reviewer: ```suggestion
1. New node creation, like adding a new column.
```

> 2. Updates for in-place node updates like changing an existing table option.

Reviewer: ```suggestion
2. In-place node updates, like changing an existing table definition.
```

> 3. Deletions for destructive node updates. Because we cannot update some nodes in- place like indexes, we should drop the original one and create the new one instead.

Reviewer: ```suggestion
3. Deletion triggered destruction node updates. Because we cannot update some nodes in place like indexes, we should drop the original one and create the new one instead.
```

> 4. Additions for destructive node updates.

Reviewer: ```suggestion
4. Addition triggered destructive node updates.
```

> 5. Deletions for deleted nodes.

Reviewer: ```suggestion
5. Node deletion.
```

> This sequence is correct in most cases.

Reviewer: this sentence seems to be redundant.

### Deparse

> Now, we get a lot of ASTs. Let’s convert them to SQLs.

Reviewer: ```suggestion
After, we get the processed ASTs, we convert them back to SQLs.
```

### Recap

1. Convert SQLs to ASTs.
2. Compare the ASTs.

> 3. Adjust diff result orders.

Reviewer: 3 and 4 needs to mention AST as well, since we are working on AST

4. Convert diff result to SQLs.

## Learn more

> You can visit [Bytebase marketing site](https://www.bytebase.com/docs/change-database/synchronize-schema) to learn more about using synchronizing schema.

Reviewer: ```suggestion
You can follow [docs](/docs/change-database/synchronize-schema) to learn more about using synchronizing schema.
```

> If you find this article helpful, you might also be interested in our product [Bytebase](https://bytebase.com/), a Database CI/CD tool for DevOps teams, built for Developers and DBAs, which is the only Database CI/CD product included by [CNCF Landscape](https://landscape.cncf.io/?selected=bytebase).

Reviewer: This sentence is only needed for articles not relevant to our product (e.g. those tagged with Industry). Not the case for this post
